Add serialization support for slots and subclass getstate/setstate

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62745
* #62744

Differential Revision: [D30113112](https://our.internmc.facebook.com/intern/diff/D30113112)

albanD committed:
Add serialization support for slots and subclass getstate/setstate

[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -103,7 +110,24 @@ def __reduce_ex__(self, proto):
+                attr_to_save = self.__class__.__slotnames__
+            else:
+                import copyreg
+                attr_to_save = copyreg._slotnames(self.__class__)
```
Should we replace this with a manual implementation here?

ezyang replied:
idk. When is slotnames available versus not? Is there a reference?

albanD replied:
It is not documented. But the CPython code actually uses it.
But I can't find any info on which versions of Python actually have that function in copyreg or not.
Should I just copy-paste that function here?

For reference, the function is here: https://github.com/python/cpython/blob/435a0334d341e5f8faed594d9f015746bb7845db/Lib/copyreg.py#L109-L158

ezyang replied:
Based on what you linked, it seems like you should just always call copyreg

albanD replied:
Yes. For some reason the cpython code re-checks the caching even though the function is doing it internally. Updated.

I'm not sure what is the criterion for using  the copyreg version vs copying?
If the function is there for all versions of python that we support, we're happy with using the internal API (and we'll fix it in future version if it gets removed)?

ezyang replied:
yeah i'd just do it this way with copyreg

albanD replied:
From looking at the history, the function has been here for 19 years and the only change in the past 14 years has been a minor bug-fix 4 years ago.
So I'm going to assume this is a pretty stable function and just use it :D

ezyang commented on torch/_tensor.py:
```diff
@@ -103,7 +110,24 @@ def __reduce_ex__(self, proto):
+            state = self.__getstate__()
+        else:
+            # Contrary to the cpython function, we unify dict and slots into a single dictionary
+            # instead of a tuple of two dictionaries
```
Why? BC? Seems better to separate them. Just make a `_rebuild_from_type_v2`

albanD replied:
Updated

ezyang commented on torch/_tensor.py:
```diff
@@ -103,7 +110,24 @@ def __reduce_ex__(self, proto):
+                attr_to_save = copyreg._slotnames(self.__class__)
+            # Tensors always have a __dict__
+            attr_to_save += ["__dict__"]
+            state = {name: getattr(self, name) for name in attr_to_save if hasattr(self, name)}
```
Is getattr/hasattr here correct semantics in the presence of `__getattr__` and friends?

albanD replied:
For the getattr:
- For the `__dict__` cpython gets it as one element [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4912) and returns it as-is [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4919). We do the same here using getattr to get the full dict in one go.
- For each slot element, cpython calls getattr() for each of them, skipping if it doesn't exit [here](https://github.com/python/cpython/blob/c83919bd635f4433f1c6ae8504996a9fe3c215e5/Objects/typeobject.c#L4969) just like we do.

For the setting:
- In cpython, all elements in the `__dict__` are set [here](https://github.com/python/cpython/blob/9ad8f109ac037ac088e09dcd2da523322c5caf34/Modules/_pickle.c#L6710) one after the other. In the code above, I only set `ret.__dict__` directly because that's what the old code was doing. Should I update it to call `setattr()` for each element in the dict?
- For other slot elements, cpython does a setattr() for each [here](https://github.com/python/cpython/blob/9ad8f109ac037ac088e09dcd2da523322c5caf34/Modules/_pickle.c#L6732) just like we do.

ezyang replied:
I guess my question was more, getattr respects `__getattr__`

```
>>> class A:
...   def __getattr__(self, x):
...     print(x)
... 
>>> A().x
x
>>> getattr(A(), 'x')
x
```

and I wasn't sure if `_PyObject_LookupAttr` also respects the magic method or not.

albanD replied:
It does respect both `tp_getattr` and `tp_getattro` which are the ones delegating to `__getattr__` or `__getattribute__ `.

ezyang commented on torch/_tensor.py:
```diff
@@ -30,12 +30,19 @@ def wrapped(*args, **kwargs):
+    ret = func(*args).as_subclass(new_type)
+    # Tensor does define __setstate__ even though it doesn't define
+    # __getstate__. So only use __setstate__ if it is NOT the one defined
+    # on Tensor
```
Why does Tensor define setstate lol

albanD replied:
I don't know!
It even states there explicitely that `torch.load()` never uses it. And since we don't have a corresponding `__getstate__`, I don't know in which case that actually can be used properly...

ezyang commented:
this seems pretty reasonable

albanD commented on torch/_tensor.py:
```diff
@@ -30,12 +30,19 @@ def wrapped(*args, **kwargs):
 
-def _rebuild_from_type(func, type, args, dict):
-    if type is Tensor:
+def _rebuild_from_type(func, new_type, args, state):
```
Ho right, I will need to keep both functions in this file though right?

albanD committed:
Update on "Add serialization support for slots and subclass getstate/setstate"




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30113112).

ezyang reviewed (approved):
okey dokey

albanD committed:
Update on "Add serialization support for slots and subclass getstate/setstate"


Differential Revision: [D30113112](https://our.internmc.facebook.com/intern/diff/D30113112)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30113112).
